### ISSUE TITLE
memory leak in asyncData on client-side

### ISSUE THREAD

**Issue by hangaoke1** [nuxt/nuxt.js#4554]
### index.vue
```vue
<template>
  <section class="container">
    <nuxt-link to="/test">test</nuxt-link>
  </section>
</template>

<script>
import Logo from '~/components/Logo.vue'

export default {
  components: {
    Logo
  },
  async asyncData() {
    return {}
  }
}
</script>
```
### test.vue
```vue
<template>
  <div>
    <h1>test</h1>
    <nuxt-link to="/">home</nuxt-link>
  </div>
</template>
```
### node_modules/@nuxt/vue-app/template/utils.js
```javascript
export function applyAsyncData(Component, asyncData) {
  const ComponentData = Component.options.data || noopData
  // Prevent calling this method for each request on SSR context
  if (!asyncData && Component.options.hasAsyncData) {
    return
  }
  console.log('mix')  // debug console
  Component.options.hasAsyncData = true
  Component.options.data = function () {
    const data = ComponentData.call(this)
    console.log(data) // debug console
    if (this.$ssrContext) {
      asyncData = this.$ssrContext.asyncData[Component.cid]
    }
    return { ...data, ...asyncData }
  }
  if (Component._Ctor && Component._Ctor.options) {
    Component._Ctor.options.data = Component.options.data
  }
}
```
/ -> /test -> / -> /test -> / -> /test ....
![image](https://user-images.githubusercontent.com/19623160/50012957-e42f1d80-fffa-11e8-9042-2e1bdedcd5c7.png)

Sometimes this is not what the user wants.

Some simple thoughts of mine
```javascript
export function applyAsyncData(Component, asyncData) {
  const ComponentData = Component.options.data || noopData
  // Prevent calling this method for each request on SSR context
  if (!asyncData && Component.options.hasAsyncData) {
    return
  }
  console.log('mix')  // debug console
  Component.options.hasAsyncData = true
  if (!Component.options.originDatafn) {
    Component.options.originDatafn = ComponentData
  }
  Component.options.data = function () {
    const data = Component.options.originDatafn.call(this)
    console.log(data) // debug console
    if (this.$ssrContext) {
      asyncData = this.$ssrContext.asyncData[Component.cid]
    }
    return { ...data, ...asyncData }
  }
  if (Component._Ctor && Component._Ctor.options) {
    Component._Ctor.options.data = Component.options.data
  }
}
```

<div align="right"><sub><em>This question is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8310">#c8310</a>)</em></sub></div>

**Comment by Anima-t3d** [comment]
I had `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory` in `asyncData`, using https://github.com/webpack/webpack/issues/6929#issuecomment-447118252 seems to fix it.

**Comment by manniL** [comment]
@hangaoke1 could you check if it happens in production mode as well?

**Comment by hangaoke1** [comment]
@manniL yes，it happens in production

**Comment by manniL** [comment]
@hangaoke1 Thanks for the info! Can you add a reproduction CodeSandbox please?

**Comment by yariksav** [comment]
Duplicate reproduce link here https://codesandbox.io/s/32m2m4j4kq
I suppose this is **critical bug**, because this do overlay datas from previous page usage

**Comment by manniL** [comment]
Resolved by #4966
